File: algorithm/단속카메라_최철원.java
import java.util.*;
import java.io.*;

class Solution {
    public class Info implements Comparable<Info>{
        int start,end;
        public Info(int start, int end){
            this.start=start;
            this.end=end;
        }
        @Override
        public int compareTo(Info o){
            return Integer.compare(this.end,o.end);
        }
    }
    public int solution(int[][] routes) {
        int answer = 0;
        PriorityQueue<Info> pq = new PriorityQueue<>();
        for(int i=0;i<routes.length;i++){
            pq.offer(new Info(routes[i][0],routes[i][1]));
        }
        Info ii = pq.poll();
        int right = ii.end;
        answer++;
        while(!pq.isEmpty()){
            ii = pq.poll();
            int cl = ii.start;
            int cr = ii.end;
            if(right<cl){
                right = cr;
                answer++;
            }
        }
        return answer;
    }
}

File: algorithm/단속카메라_최철원.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>
#include <iostream>
using namespace std;
struct info{
    int start,end;
};
struct cmp{
    bool operator()(info &a, info &b){
        if(a.end==b.end){
            return a.start > b.start;
        }
        else return a.end > b.end;
    }
};

int solution(vector<vector<int>> routes) {
    int answer = 0;
    priority_queue<info,vector<info>,cmp> pq;
    for(int i=0;i<routes.size();i++){
        pq.push({routes[i][0],routes[i][1]});
    }
    answer++;
    int left = pq.top().start;
    int right = pq.top().end;
    pq.pop();
    while(!pq.empty()){
        int cl = pq.top().start;
        int cr = pq.top().end;
        pq.pop();
        if(right<cl){       //새로운 단속
            answer++;
            right = cr;
        }
        /* 뒤쪽 오름차순 정렬 + left는 비교대상이 아니기 때문에 필요없음
        else{       //겹칠 때
            left = max(left,cl);
            right = min(right,cr);
        }*/
    }
    return answer;
}
